addon/lib/field.js

```diff
@@ -207,6 +208,130 @@ export default Base.extend({
+   *
+   * @method _fetchUsedDynamicOptions.perform
+   * @return {Object[]} Formerly used dynamic options
+   * @private
+   */
+  _fetchUsedDynamicOptions: task(function*() {
+    if (!this.question.isDynamic) return null;
+
+    const edges = yield this.apollo.query(
+      {
+        query: getDocumentUsedDynamicOptionsQuery,
+        fetchPolicy: "cache-first",
+        variables: {
+          document: this.document.uuid,
+          question: this.question.slug
+        }
+      },
+      "allUsedDynamicOptions.edges"
+    );
+
+    return edges.map(({ node: { slug, label } }) => ({
+      slug,
+      label
+    }));
+  }),
+
+  /**
+   * The formerly used dynamic options for this question.
+   *
+   * @property {Object[]} usedDynamicOptions
+   * @accessor
+   *
+   */
+  usedDynamicOptions: reads("_fetchUsedDynamicOptions.lastSuccessful.value"),
+
+  /**
+   * The available options for choice questions. This only works for the
+   * following types:
+   *
+   * - ChoiceQuestion
+   * - DynamicChoiceQuestion
+   * - MultipleChoiceQuestion
+   * - DynamicMultipleChoiceQuestion
+   *
+   * This will also return the disabled state of the option. An option can only
+   * be disabled, if it is an old value used in a dynamic question.
+   *
+   * @property {Null|Object[]} options
+   * @accessor
+   */
+  options: computed(
+    "value",
+    "question.options.[]",
+    "usedDynamicOptions.[]",
+    function() {
+      if (!this.question.isChoice && !this.question.isMultipleChoice) {
+        return null;
+      }
+
+      const selected =
+        (this.question.isMultipleChoice ? this.value : [this.value]) || [];
+
+      const options = this.question.options.map(option => ({
+        ...option,
+        disabled: false
+      }));
+
+      const hasInvalidSelected = !selected.every(slug =>
+        options.find(option => option.slug === slug)
+      );
+
+      if (this.question.isDynamic && hasInvalidSelected) {
+        if (!this._fetchUsedDynamicOptions.lastSuccessful) {
+          // Fetch used dynamic options if not done yet
+          this._fetchUsedDynamicOptions.perform();
+        }
+
+        return [
+          ...options,
+          ...(this.usedDynamicOptions || [])
+            .filter(used => {
+              return (
+                selected.includes(used.slug) &&
+                !options.find(option => option.slug === used.slug)
+              );
+            })
+            .map(used => ({ ...used, disabled: true }))
+        ];
+      }
+
+      return options;
+    }
+  ),
+
+  /**
+   * The currently selected option. This property is only used for choice
+   * questions. It can either return null if no value is selected yet, an
+   * object for single choices or an array of objects for multiple choices.
+   *
+   * @property {Null|Object|Object[]} selected
+   * @accessor
+   */
+  selected: computed(
+    "value",
+    "options.@each.slug",
+    "usedDynamicOptions.@each.slug",
```

Comment:
Shouldn't this be `options.[]` and `usedDynamicOptions.[]`?

If I understand the rest of the code correctly, the slugs themselves don't change.

Reply:
The only property of those options we need is `slug` so it makes no sense to observe the whole array

Reply:
However, `"usedDynamicOptions.@each.slug"` is now obsolete, will remove

Reply:
Ok